core/src/main/java/com/graphhopper/routing/QueryGraph.java

```diff
@@ -356,6 +352,98 @@ private void createEdges( GHPoint3D prevSnapped, int prevWayIndex, GHPoint3D cur
         virtualEdges.add(baseReverseEdge);
     }
 
+    /**
+     * set edges at queryResult disprefered which require at least a turn of 100° from preferredDirectionAz
+     * @param preferredDirectionAz north based Azimuth
+     * @param incoming if true, incoming edges are penalized, else outgoing edges
+     */
+    public void enforceDirection(Double preferredDirectionAz, QueryResult queryResult, boolean incoming)
+    {
+        double preferredDirection;
+        if (!isInitialized())
+        {
+            throw new IllegalStateException("QueryGraph.lookup has to be called in before direction enforcement");
+        }
+        if (Double.isNaN(preferredDirectionAz))
+        {
+            return;
+        } else
+        {
+            preferredDirection = ac.convertAzimuth2xaxisAngle(preferredDirectionAz);
+        }
+        
+        int nodeId =  queryResult.getClosestNode();
+        if (!isVirtualNode(nodeId))
+        {
+            throw new IllegalStateException("Direction enforcement should only happen at virtual nodes");
+        }
+        int virtNodeIDintern = nodeId - mainNodes;
+        
+        // either penalize incoming or outgoing edges
+        List<Integer> edgePositions = incoming? Arrays.asList(VE_BASE, VE_ADJ_REV) : Arrays.asList(VE_BASE_REV, VE_ADJ);
+        for (int edgePos : edgePositions)
+        {
+            VirtualEdgeIteratorState edge = virtualEdges.get(virtNodeIDintern * 4 + edgePos);
+
+            PointList wayGeo = edge.fetchWayGeometry(3);
+            double edgeOrientation;
+            if (incoming)
+            {
+                int numWayPoints = wayGeo.getSize();
+                edgeOrientation = ac.calcOrientation(wayGeo.getLat(numWayPoints-2), wayGeo.getLon(numWayPoints-2),
+                        wayGeo.getLat(numWayPoints-1), wayGeo.getLon(numWayPoints-1));
+            } else
+            {
+                edgeOrientation =ac.calcOrientation(wayGeo.getLat(0), wayGeo.getLon(0),
+                    wayGeo.getLat(1), wayGeo.getLon(1));
+            }
+            
+            edgeOrientation = ac.alignOrientation(preferredDirection, edgeOrientation);
+            double delta = (edgeOrientation - preferredDirection);
+
+            if (Math.abs(delta) > 1.74) // penalize if a turn of more than 100° is necessary
+            {
+                edge.setDispreferedEdge(true, false);
+                //also apply to opposite edge for reverse routing
+                VirtualEdgeIteratorState reverseEdge = virtualEdges.get(virtNodeIDintern * 4 + getPosOfReverseEdge(edgePos));
+                reverseEdge.setDispreferedEdge(true, true);
+            }
+        }
+    }
+
+    /**
+     * set edgeId disprefered (in direction adjNodeId)
+     */
+    public void setDispreferedEdge(int edgeId, int adjNodeId)
```

Comment:
we need to find a better name (also prefer**r**ed) or maybe setInferiorEdge? And the sout statements should be removed. And do we really need to make this low level method public?